charmhub/client.go

```diff
@@ -19,170 +19,66 @@ import (
 type Logger interface {
 	IsTraceEnabled() bool
 
 	Errorf(string, ...interface{})
-	Debugf(string, ...interface{})
 	Tracef(string, ...interface{})
-
-	ChildWithLabels(string, ...string) loggo.Logger
 }
 
 // Config holds configuration for creating a new charm hub client.
+// The zero value is a valid default configuration.
 type Config struct {
-	// URL holds the base endpoint URL of the charmHub,
+	// URL holds the base endpoint URL of the Charmhub API,
 	// with no trailing slash, not including the version.
-	// For example https://api.charmhub.io/v2/charms/
+	// If empty string, use the default Charmhub API server.
 	URL string
 
-	// Version holds the version attribute of the charmHub we're requesting.
-	Version string
-
-	// Entity holds the entity to target when querying the API (charm or snaps).
-	Entity string
-
-	// Headers allow the defining of a set of default headers when sending the
-	// requests. These headers augment the headers required for sending requests
-	// and allow overriding existing headers.
-	Headers http.Header
-
-	// Transport represents the default http transport to use for all API
-	// requests.
-	Transport Transport
+	// HTTPClient represents the HTTP client to use for all API
+	// requests. If nil, use the default HTTP client.
+	HTTPClient HTTPClient
 
-	// Logger to use during the API requests.
+	// Logger to use during the API requests. If nil, don't log anything.
 	Logger Logger
-}
-
-// Option to be passed into charmhub construction to customize the client.
-type Option func(*options)
-
-type options struct {
-	url             *string
-	metadataHeaders map[string]string
-	transportFunc   func(Logger) Transport
-}
 
-// WithURL sets the url on the option.
-func WithURL(u string) Option {
-	return func(options *options) {
-		options.url = &u
-	}
+	// FileSystem represents the file system operations for downloading.
+	// If nil, use the real OS file system.
+	FileSystem FileSystem
 }
 
-// WithMetadataHeaders sets the headers on the option.
-func WithMetadataHeaders(h map[string]string) Option {
-	return func(options *options) {
-		options.metadataHeaders = h
-	}
-}
-
-// WithHTTPTransport sets the default http transport to use on the option.
-func WithHTTPTransport(transportFn func(Logger) Transport) Option {
-	return func(options *options) {
-		options.transportFunc = transportFn
-	}
-}
-
-// Create a options instance with default values.
-func newOptions() *options {
-	u := CharmHubServerURL
-	return &options{
-		url: &u,
-		transportFunc: func(logger Logger) Transport {
-			return DefaultHTTPTransport(logger)
-		},
-	}
-}
-
-// CharmHubConfig defines a charmHub client configuration for targeting the
-// charmhub API.
-func CharmHubConfig(logger Logger, options ...Option) (Config, error) {
-	opts := newOptions()
-	for _, option := range options {
-		option(opts)
-	}
-
-	if logger == nil {
-		return Config{}, errors.NotValidf("nil logger")
-	}
```

Comment:
Should not be removed, providing a valid logger is required. We've had to rely on the data provided too many times already.

Reply:
Done.